File: problem_016.hs
-- 2^(15) = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.
-- 
-- What is the sum of the digits of the number 2^(1000)?
-- 
-- http://projecteuler.net/index.php?section=problems&id=16

module Main where

digits :: Integral t => t -> [t]
digits 0 = []
digits n = (mod n 10) : digits (div n 10)

main :: IO ()
main = do print $ sum $ digits $ 2 ^ 1000

File: c/problem_016.c
/*
 * Problem 016
 *
 * × = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.
 *
 * What is the sum of the digits of the number 2^1000?
 *
 * https://projecteuler.net/problem=16
 */

#include <euler.h>

int main(void) {
  mpz_t two;

  mpz_init_set_ui(two, 2);

  mpz_pow_ui(two, two, 1000);

  printf("%lu\n", mpz_sum_of_digits(two));
}
